packages/core/src/provider/BaseProvider.ts

```diff
@@ -2,19 +2,280 @@ import { ComposioGlobalExecuteToolFnNotSetError } from '../errors/ToolErrors';
 import { ExecuteToolModifiers } from '../types/modifiers.types';
 import type { Tool, ToolExecuteParams, ToolExecuteResponse } from '../types/tool.types';
 import { ExecuteToolFn, GlobalExecuteToolFn } from '../types/provider.types';
+import { McpCreateResponse } from '@composio/client/resources/index';
+import ComposioClient from '@composio/client';
+import {
+  CustomCreateResponse,
+  McpListResponse,
+  McpUpdateResponse,
+  McpDeleteResponse,
+} from '@composio/client/resources/mcp';
+import { MCPToolkitConfig, MCPAuthOptions, MCPGenerateURLParams, MCPGetServerParams } from '../types/mcp.types';
+
+export type McpServerUrlInfo = {
+  url: URL;
+  name: string;
+  toolkit?: string;
+};
+
+export type McpServerGetResponse = McpServerUrlInfo | McpServerUrlInfo[];
+
+export type McpUrlResponse = {
+  connected_account_urls?: string[];
+  user_ids_url?: string[];
+  mcp_url: string;
+};
+
+export type McpServerCreateResponse<T> = (McpCreateResponse | CustomCreateResponse) & {
+  toolkits: string[];
+  getServer: (params: MCPGetServerParams) => Promise<T>;
+};
+
+export abstract class McpProvider<T> {
+  client?: ComposioClient;
+
+  setup(client: ComposioClient): void {
```

Comment:
Exposing the client to providers is not good as it exposes unecessary scopes to the provider implementation. Ideally only expose the required methods here.